src/ddsrt/include/dds/ddsrt/events.h

```diff
@@ -0,0 +1,179 @@
+//#define DDSRT_EVENT_FLAG_IP_CHANGE (1u<<5)  
+///@}
+
+  /**
+  * @brief Describes the type of object being monitored for events.
+  */
+  enum ddsrt_event_type {
+    ddsrt_event_type_unset, /**< unitialized state*/
+    ddsrt_event_type_socket /**< indicating a socket type connection*/
+    /*,
+    ddsrt_event_type_file,
+    ddsrt_event_type_ifaddr*/
+  };
+  typedef enum ddsrt_event_type ddsrt_event_type_t;
+
+  /**
+  * @brief Structure containing the information of which object is being watched.
+  *
+  * Is used by ddsrt_event_queue to keep track of objects being watched.
+  */
+  struct ddsrt_event {
+    uint32_t flags; /**< Type of event being watched for, composited from DDSRT_EVENT_FLAG_$EVENT_TYPE$.*/
+    ddsrt_event_type_t type; /**< Type of object being watched.*/
+    ddsrt_atomic_uint32_t triggered; /**< Trigger status after a call to ddsrt_event_queue_wait.*/
+    union {
+      struct { ddsrt_socket_t sock; } socket;
+      /*struct ddsrt_event_file file;  this is for future expansion to be able to register events on files*/
+      /*struct ddsrt_event_ifaddr ifaddr; this is for future expansion to be able to register IP address changes*/
+    } data; /**< Container for the object being watched.*/
+  };
+
+  typedef struct ddsrt_event ddsrt_event_t;
+
+  typedef struct ddsrt_event_queue ddsrt_event_queue_t;
+  
+  /**
+  * @brief Initializes an existing event to indicate a socket.
+  *
+  * @param[out] ev Pointer to the event to initialize.
+  * @param[in] sock Socket to initialize the event with.
+  * @param[in] flags Flags to set for the event.
+  *
+  * @returns DDS_RETCODE_OK if everything went OK.
+  */
+  dds_return_t ddsrt_event_socket_init(ddsrt_event_t* ev, ddsrt_socket_t sock, uint32_t flags);
+
+  /**
+  * @brief Event queue creation function.
+  *
+  * @returns Pointer to the created queue.
+  */
+  ddsrt_event_queue_t* ddsrt_event_queue_create(void);
+
+  /**
+  * @brief Cleans up an event queue.
+  *
+  * Will finish the event queue first, then free the memory of the queue.
+  *
+  * @param[out] queue The queue to clean up.
+  *
+  * @returns DDS_RETCODE_OK if everything went OK.
+  */
+  dds_return_t ddsrt_event_queue_destroy(ddsrt_event_queue_t* queue);
+
+  /**
+  * @brief Getter for the number of stored events.
+  *
+  * @param[in] queue The queue to get the number of events of.
+  *
+  * @returns The number of events stored by the queue.
+  */
+  size_t ddsrt_event_queue_nevents(ddsrt_event_queue_t* queue);
+
+  /**
+  * @brief Triggers a wait for events for the queue.
+  *
+  * @param[out] queue The queue to trigger.
+  * @param[in] reltime The maximum amount of time to wait.
+  *
+  * @returns DDS_RETCODE_OK if everything went OK.
+  */
+  dds_return_t ddsrt_event_queue_wait(ddsrt_event_queue_t* queue, dds_duration_t reltime);
+
+  /**
+  * @brief Interrupts a triggered wait of a queue.
+  *
+  * @param[out] queue The queue to interrupt.
+  *
+  * @returns DDS_RETCODE_OK if everything went OK.
```

Comment:
Fully agree with this, I will improve the documentation to describe all returned values and under which conditions (if it can be described within a sentence or two).
About the choice for returning errors or calling an abort, I will go over the code where errors are returned and see whether the state of the program is recoverable, if it is not, i will use some kind of hard stop.

Reply:
When it comes to nitpicking `@retval` should be used for this sort of thing. `@returns` describes what is returned.

Reply:
I will also have a look which functions really need to have a return value, i.e. functions which can only return a single value.

Reply:
This is now implemented.